app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,197 @@
+    end
 
-  def build_batch
-    batch = Sidekiq::Batch.new
-    batch.description = batch_description
-    batch_callbacks(batch) { batch.jobs { destroy_and_delete_associations } }
-    batch
-  end
+    private
 
-  def batch_description
-    "Deleting #{object.class.name} [##{object.id}]"
-  end
+    # @return the hierarchy entries to handle deletion of an object
+    def hierarchy_entries_for(ar_object)
+      return [] unless ar_object&.persisted? # e.g. when calling Proxy#oidc_configuration a new object can be generated
+
+      if ar_object.respond_to?(:destroyable?, true) && !ar_object.send(:destroyable?)
+        raise DoNotRetryError, "Background deleting #{ar_object.class}:#{ar_object.id} which is not destroyable."
+      elsif ar_object.is_a?(FeaturesPlan)
+        # This is an ugly hack to handle lack of `#id` but we have only FeaturesPlans with a composite primary key.
+        # Rails 7.1 supports composite primary keys so we can implement universal handling. See [FeaturesPlan].
+        # Now to avoid complications, just sweep it under the rag.
+        FeaturesPlan.where(feature_id: ar_object.feature_id, plan_id: ar_object.plan_id).delete_all
+        return []
+      end
 
-  def batch_callbacks(batch)
-    %i[success complete].each { |name| batch.on(name, self.class, callback_options) }
-    yield
-    bid = batch.bid
+      ar_object_str = "#{ar_object.class}-#{ar_object.id}"
 
-    if Sidekiq::Batch::Status.new(bid).total.zero?
-      on_complete(bid, callback_options)
-    else
-      info("DeleteObjectHierarchyWorker#batch_success_callback retry job with the hierarchy of workers: #{caller_worker_hierarchy}")
-      retry_job wait: 5.minutes
+      [
+        "Plain-#{ar_object_str}",
+        *ar_object.background_deletion.map { "Association-#{ar_object_str}:#{_1}" },
+      ]
     end
   end
 
-  def destroy_and_delete_associations
-    Array(object.background_deletion).each do |association_config|
-      reflection = BackgroundDeletion::Reflection.new(association_config)
-      next unless destroyable_association?(reflection.name)
+  # @param hierarchy [Array<String>] something like ["Plain-Service-1234", "Association-Service-1234:plans" ...]
+  # @note processes deleting a hierarchy of objects and reschedules itself at current progress after a time limit
+  def perform(*hierarchy)
+    return compatibility(*hierarchy) unless hierarchy.first.is_a?(String)
 
-      ReflectionDestroyer.new(object, reflection, caller_worker_hierarchy).destroy_later
+    started = now
+    while hierarchy.present? && now - started < WORK_TIME_LIMIT_SECONDS
+      Rails.logger.info "Starting background deletion iteration with: #{hierarchy.join(' ')}"
+      handle_one_hierarchy_entry!(hierarchy)
     end
+
+    @remaining_hierarchy = hierarchy
   end
 
-  def destroyable_association?(_association)
-    true
+  # we can just use job.arguments.first but for compatibility mode we want it not to lock
```

Comment:
```suggestion
  # Custom lock key for `activejob-uniqueness`
  # we can just use job.arguments.first but for compatibility mode we want it not to lock
```

Reply:
Actually, after we remove compatibility mode - cannot we just use the default lock key, without overriding?

Reply:
Even if there is a default one (perhaps one comparing all arguments), no. The lock key needs to effectively prevent duplicate jobs. And during rescheduling the current hierarchy may change. So we want the one unique thing about the job.

Reply:
Yeah, right... I still need to interiorize the implementation, because I keep confusing it with the old one.

Reply:
And this :point_up: